dynamiq/nodes/agents/orchestrators/graph.py

```diff
@@ -0,0 +1,531 @@
+
+        result = context.pop("result")
+
+        return result, context
+
+    def execute(self, input_data: StateInputSchema, config: RunnableConfig = None, **kwargs) -> dict:
+        logger.debug(f"State {self.id}: starting the flow with input_task:\n```{input_data}```")
+        kwargs.pop("run_depends", None)
+        self.run_on_node_execute_run(config.callbacks, **kwargs)
+        kwargs = kwargs | {"parent_run_id": kwargs.get("run_id")}
+
+        global_context = input_data.context
+        chat_history = input_data.chat_history
+
+        if len(self.tasks) == 1:
+            result, context = self._submit_task(
+                self.tasks[0],
+                global_context,
+                chat_history,
+                config=config,
+                **kwargs,
+            )
+
+            chat_history.append(
+                {
+                    "role": "system",
+                    "content": f"Result: {result}",
+                }
+            )
+
+            global_context = global_context | context
+
+        elif len(self.tasks) > 1:
+
+            contexts = []
+
+            with ThreadPoolExecutor() as executor:
+
+                futures = [
+                    executor.submit(
+                        self._submit_task,
+                        task,
+                        copy.deepcopy(global_context),
+                        copy.deepcopy(chat_history),
+                        config=config,
+                        **kwargs,
+                    )
+                    for task in self.tasks
+                ]
+                for future in futures:
+
+                    result, context = future.result()
+                    chat_history.append(
+                        {
+                            "role": "system",
+                            "content": f"Result: {result}",
+                        }
+                    )
+
+                    contexts.append(context)
+            global_context = global_context | self.merge_contexts(contexts)
+
+        return {"context": global_context, "chat_history": chat_history}
+
+
+START = "START"
+END = "END"
+
+
+class GraphOrchestrator(Orchestrator):
+    """
+    Orchestrates the execution of complex tasks using multiple specialized agents.
+
+    This class manages the breakdown of a main objective into subtasks,
+    delegates these subtasks to appropriate agents, and synthesizes the results
+    into a final answer.
+
+    Attributes:
+        manager (ManagerAgent): The managing agent responsible for overseeing the orchestration process.
+        agents (List[BaseAgent]): List of specialized agents available for task execution.
+        objective (Optional[str]): The main objective of the orchestration.
+        max_loops (Optional[int]): Maximum number of transition between states.
+    """
+
+    name: str | None = "GraphOrchestrator"
+    manager: GraphAgentManager
+    initial_state: str = START
+
+    states: dict[str, State] = {}
```

Comment:
Lets use list of `states` similar to other usages: agent/state `tasks`,  flow `nodes`, workflow `flows` and so on.
You could create `_state_by_id` on init for internal usage (similar to Flow)